DM Hackathon- Add a daily prompt modal to editor.

Addison-Stavlo opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to # DM Hackathon Project

https://github.com/Automattic/wp-calypso/pull/84240 adds query args to trigger this within the blogger onboarding flow.

## Proposed Changes
DM HACKATHON PROJECT

* adds a modal to allow users to select a daily prompt. adds the daily prompt block when selected. modal opens when `new_prompt` query arg is present.
* adds tracking events for `calypso_editor_writing_prompts_modal_viewed`, `calypso_editor_writing_prompts_modal_prompt_selected` (with a `promptId` prop), and `calypso_editor_writing_prompts_modal_closed`.

![prompt-modal-2-small](https://github.com/Automattic/wp-calypso/assets/28742426/311aafec-db77-44ba-9bbd-2da801797b3c)


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Setup:
* Sandbox a test site, public-api, and widgets.wp.com
* build and sync wpcom-block-editor from this branch ( `yarn dev --sync` from apps/wpcom-block-editor)
* build and sync editing-toolkit from this branch (`yarn dev --sync` from apps/editing-toolkit)
* build calypso from this branch

Test the Modal:
* go to the path `*/post/{siteSlug}/?new_prompt=true` to open the editor with this arg present and the modal should appear to browse and select prompts.
* test tracking events for the modal - `calypso_editor_writing_prompts_modal_viewed`, `calypso_editor_writing_prompts_modal_prompt_selected` (with a `promptId` prop), and `calypso_editor_writing_prompts_modal_closed` - and verify they fire when expected.

Test the writing prompts card on My Home:
* go to my home on local calypso for a blogging site where the daily prompts card is available.
* select a writing prompt and select "Post Answer"
* verify that the editor loads with the writing prompt block for this prompt in place.

Smoke check entering the editor from other sources and without these query params.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Addison-Stavlo committed:
piggy-back off press this as a starting point to insert prompt block for daily prompt entry points

Addison-Stavlo committed:
use default promptId if answer_prompt is strictly true

Addison-Stavlo committed:
support new_prompt query arg to signal inserting block without a specific answer_prompt id

Addison-Stavlo committed:
refactor to be separate from press-this

Addison-Stavlo committed:
try a prompt selection modal, initial, needs style and love

Addison-Stavlo commented on apps/wpcom-block-editor/src/default/features/blogging-prompt.js:
```diff
@@ -0,0 +1,18 @@
+	( async () => {
+		await isEditorReady();
+
+		const blockProps = answer_prompt ? { promptId: answer_prompt } : {};
```
noting this is unnecessary now that we took `new_prompt` out of this functionality to use with the modal instead. we can always use `promptId` with this value.

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/blogging-prompts-modal/index.js:
```diff
@@ -0,0 +1,63 @@
+			.then( ( result ) => {
+				return setPrompts( result );
+			} )
+			.catch( ( err ) => setError( err ) );
```
noting we arent doing anything with the error anyways.

sixhours committed:
Style this puppy

sixhours committed:
whitespace cleanup

Addison-Stavlo committed:
refactor and avoid rendering component when unnecessary

Addison-Stavlo committed:
add tracks events

Addison-Stavlo committed:
Merge branch 'trunk' into try/daily-prompt-modal

Addison-Stavlo committed:
use console.error instead of console.log for catch response

roo2 commented:
there seems to be a visual issue with the mobile version
<img width="501" alt="Screen Shot 2023-12-07 at 10 04 40 am" src="https://github.com/Automattic/wp-calypso/assets/22446385/d5c23c9c-7582-4ef2-9cb5-bb93c40f643d">

I think that on tablet sized screens, we could increase the width of the text a bit to avoid too much wrapping
<img width="1006" alt="Screen Shot 2023-12-07 at 10 06 29 am" src="https://github.com/Automattic/wp-calypso/assets/22446385/95939920-7a17-4a53-9101-694cfe21ad74">

roo2 commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/blogging-prompts-modal/index.js:
```diff
@@ -0,0 +1,102 @@
+
+export const BloggingPromptsModalInner = () => {
+	const translate = useTranslate();
+	const [ isOpen, setIsOpen ] = useState( true );
```
I think we should prevent it from opening for pages or when editing existing posts, it should only be available for new posts.

Addison-Stavlo replied:
Good point, we can add some optional checks. But really its only opened with the presence of the query arg, which we control. There should be no circumstances where this arg is added to the editor for pages or editing posts, unless we or the user specifically desired to add those manually.

roo2 commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/blogging-prompts-modal/index.js:
```diff
@@ -0,0 +1,102 @@
+import { recordTracksEvent } from '@automattic/calypso-analytics';
```
nit: It could be done in typescript since it's new code

Addison-Stavlo replied:
I was thinking about this, but Im curious what the benefit would be in this circumstance:
* the component does not take any props
* the component is not connected to any external data in the codebase
* the component is not intended for reuse, has no children, and seems very self-contained.

With that in mind, what would typescript buy us here? It doesn't seem like there is any potential for it to catch regressions due to other changes in the codebase. It seems like it would just add weight to reading and updating with no benefit. I could be wrong though, but what would the benefits be in this case?

roo2 replied:
yea probably not very much! I was just thinking as a default, but I don't think it's particularly useful here either

Addison-Stavlo replied:
Ugg, i tried this briefly and what a PITA. The TS linter gets (inconsistently) mad at things like the window object, using hook setters with values returned from the fetch, using dispatch from wp-data in any way... and the fixes others seemed to have used in these areas feel like hacky work arounds (like making placeholder types simply to satisfy the linter locally). Since we didn't find how converting this to TS would be beneficial given its encapsulation in the first place, these silly problems it surfaces, etc. It really doesn't seem valuable to spend time on this.

Addison-Stavlo committed:
rebase trunk

Addison-Stavlo committed:
update styles for device width

Addison-Stavlo committed:
restrict to post editor

Addison-Stavlo commented:
I fixed the issues with mobile layout and added an extra check to ensure we are actually in the post editor. This should be ready for another round of review/testing.

davemart-in commented:
Tested this. A couple of notes:

1) I made a minor update to the header font style to put the emphasis on the prompt:

**Before**
![CleanShot 2024-01-04 at 13 32 35@2x](https://github.com/Automattic/wp-calypso/assets/5634774/34a24e88-b6b7-4871-94ae-bd871073838b)

**After**
<img width="826" alt="CleanShot 2024-01-04 at 13 33 44@2x" src="https://github.com/Automattic/wp-calypso/assets/5634774/66b4e4aa-f326-453c-8d74-d90bb229851c">

2) For some reason this whole UI was REALLLY slow to load. Guessing that is just a function of sandboxing and syncing various things...

3) Just noting that the new Jetpack AI prompts did not work. Not sure that is a blocker since they're A18n only at the moment, but figured I'd call it out.

Addison-Stavlo commented:
> I made a minor update to the header font style to put the emphasis on the prompt:

Thank you! Much appreciated 😁 

> For some reason this whole UI was REALLLY slow to load. Guessing that is just a function of sandboxing and syncing various things...

I think so. This includes loading the editor itself as well, and not just a large delay between loading the editor fully and having the modal appear?

> Just noting that the new Jetpack AI prompts did not work. Not sure that is a blocker since they're A18n only at the moment, but figured I'd call it out.

I think that is expected here. Those prompts do not come from the same endpoint, and where they are being tested on myHome are being retrieved from a separate endpoint and filtered in on the front-end IIRC. With that, it is also probably a good thing since those AI prompts rely on the user having written and published posts for the AI to make predictions based off of. At the point of early onboarding, they probably don't make sense for that reason. 😁

davemart-in reviewed (approved):
LGTM! Nice job getting this over the finish line! Kudos to everyone who contributed!

Addison-Stavlo committed:
add import for font size variable

Addison-Stavlo committed:
recommit daves additions to font styles

Addison-Stavlo commented:
The etk build was failing due to our font variable addition, so I added the the package import for it. Now E2Es are consistently failing on accessing revisions in the editor, so I will need to investigate. Those didn't fail on previous commits so it seems unlikely due to our changes but they are close in flow to what we are working on here so we need to be cautious...

Addison-Stavlo commented:
Ok, the 'view revisions' test failures seem like a red herring. They are now passing with no changes to the code here (note they were also failing on [this other branch](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Calypso_E2E_Playwright_desktop/11570713?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildTestsSection=true&expandBuildChangesSection=true) that got merged yesterday), and of course another test started failing here with no other changes  here... yay E2E flakynesses.

this should be good to go.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/10751351

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @Addison-Stavlo for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10751351).
